invisible_cities/cities/esmeralda.py

```diff
@@ -186,63 +164,78 @@ def get_track_energy(track):
-            list_of_vars = [hitc.event, tID, energy, length, numb_of_voxels,
-                            numb_of_hits, numb_of_tracks,
-                            min_x, min_y, min_z, max_x, max_y, max_z, max_r,
-                            ave_pos[0], ave_pos[1], ave_pos[2],
-                            extr1_pos[0], extr1_pos[1], extr1_pos[2],
-                            extr2_pos[0], extr2_pos[1], extr2_pos[2],
-                            blob_pos1[0], blob_pos1[1], blob_pos1[2],
-                            blob_pos2[0], blob_pos2[1], blob_pos2[2],
-                            e_blob1, e_blob2, overlap,
-                            vox_size_x, vox_size_y, vox_size_z]
-
-            df.loc[c] = list_of_vars
-            try:
-                types_dict
-            except NameError:
-                types_dict = dict(zip(df.columns, [type(x) for x in list_of_vars]))
-
-            for vox in t.nodes():
-                for hit in vox.hits:
-                    hit.track_id = tID
-                    track_hits.append(hit)
-
-
-        track_hitc = evm.HitCollection(hitc.event, hitc.time)
-        track_hitc.hits = track_hits
-        #change dtype of columns to match type of variables
-        df = df.apply(lambda x : x.astype(types_dict[x.name]))
 
-        return df, track_hitc
+        if len(hitc.hits)>0:
+            voxels     = plf.voxelize_hits(hitc.hits, vox_size, strict_vox_size, energy_type)
+            mod_voxels = plf.drop_end_point_voxels(voxels, energy_threshold, min_voxels)
+            tracks     = plf.make_track_graphs(mod_voxels)
+
+            vox_size_x = voxels[0].size[0]
+            vox_size_y = voxels[0].size[1]
+            vox_size_z = voxels[0].size[2]
+            del(voxels)
+            def get_track_energy(track):
+                return sum([vox.Ehits for vox in track.nodes()])
+            #sort tracks in energy
+            tracks     = sorted(tracks, key = get_track_energy, reverse = True)
+
+            track_hits = []
+            for c, t in enumerate(tracks, 0):
+                tID = c
+                energy = get_track_energy(t)
+                length = plf.length(t)
+                numb_of_hits = len([h for vox in t.nodes() for h in vox.hits])
+                numb_of_voxels = len(t.nodes())
+                numb_of_tracks = len(tracks   )
+
+                min_x = min([h.X for v in t.nodes() for h in v.hits])
+                max_x = max([h.X for v in t.nodes() for h in v.hits])
+                min_y = min([h.Y for v in t.nodes() for h in v.hits])
+                max_y = max([h.Y for v in t.nodes() for h in v.hits])
+                min_z = min([h.Z for v in t.nodes() for h in v.hits])
+                max_z = max([h.Z for v in t.nodes() for h in v.hits])
+                max_r = max([np.sqrt(h.X*h.X + h.Y*h.Y) for v in t.nodes() for h in v.hits])
+
+                pos = [h.pos for v in t.nodes() for h in v.hits]
```

Comment:
> haven't time-profiled it, but my feeling is that it is very inefficient.

The most important rule of optimization: don't optimize anything until you've proven that it matters. That means that you must either prove (by timing) that 

a) this is a bottleneck
b) the alternative implementation is actually faster that the original.

If it's not a bottleneck, but the new version is faster and not significantly uglier, then go ahead. If the new version is uglier and more difficult to understand, then don't use it **unless profiling proves that the change makes a significant improvement in perfomance**.

However, the square brackets inside the `max` and `min` ensure that these will run in linear rather than constant space, which is just a pointless waste ... so get rid of the `[]`s. That is

    fn([ for ... in ... ])

should *always* be written as

    fn( for ... in ... )

or

   fn(( for ... in ... ))

if you like pointless characters (or there are other arguments besides the listcomp / genexpr).

Reply:
> Yes, I agree, but doesn't it grind your gears to see those loops repeated
so many times?

It does, it does. A smart enough compiler ...

Reply:
I agree! However I realized that we are already constructing pos list to use it in  np.average so I just rearranged the order to first construct pos list and then look for appropriate min/max